flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/FlinkCalciteCatalogReader.java

```diff
@@ -62,17 +80,181 @@ public FlinkCalciteCatalogReader(
+					"without source.");
+			}
+		} else if (table instanceof CatalogTable) {
+			CatalogTable catalogTable = (CatalogTable) table;
+			return convertCatalogTable(relOptSchema,
+				names,
+				rowType,
+				catalogTable,
+				schemaTable);
+		} else {
+			throw new ValidationException("Unsupported table type: " + table);
+		}
+	}
+
+	private static FlinkPreparingTableBase convertTemporaryTable(
+			RelOptSchema relOptSchema,
+			List<String> names,
+			RelDataType rowType,
+			CatalogSchemaTable schemaTable) {
+		final CatalogBaseTable baseTable = schemaTable.getCatalogTable();
+		if (baseTable instanceof QueryOperationCatalogView) {
+			return convertQueryOperationView(relOptSchema,
+				names,
+				rowType,
+				(QueryOperationCatalogView) baseTable);
+		} else if (baseTable instanceof ConnectorCatalogTable) {
+			ConnectorCatalogTable<?, ?> connectorTable = (ConnectorCatalogTable<?, ?>) baseTable;
+			if ((connectorTable).getTableSource().isPresent()) {
+				return convertSourceTable(relOptSchema,
+					names,
+					rowType,
+					connectorTable,
+					schemaTable.getStatistic(),
+					schemaTable.isStreamingMode());
+			} else {
+				throw new ValidationException("Cannot convert a connector table " +
+					"without source.");
+			}
+		} else if (baseTable instanceof CatalogTable) {
+			return convertCatalogTable(relOptSchema,
+				names,
+				rowType,
+				(CatalogTable) baseTable,
+				schemaTable);
+		} else {
+			throw new ValidationException("Unsupported table type: " + baseTable);
+		}
+	}
+
+	private static FlinkPreparingTableBase convertQueryOperationView(
+			RelOptSchema relOptSchema,
+			List<String> names,
+			RelDataType rowType,
+			QueryOperationCatalogView view) {
+		return QueryOperationCatalogViewTable.create(relOptSchema, names, rowType, view);
+	}
+
+	private static FlinkPreparingTableBase convertSourceTable(
+			RelOptSchema relOptSchema,
+			List<String> names,
+			RelDataType rowType,
+			ConnectorCatalogTable<?, ?> table,
+			FlinkStatistic statistic,
+			boolean isStreamingMode) {
+		TableSource<?> tableSource = table.getTableSource().get();
+		if (!(tableSource instanceof StreamTableSource ||
+			tableSource instanceof LookupableTableSource)) {
+			throw new ValidationException(
+				"Only StreamTableSource and LookupableTableSource can be used in Blink planner.");
+		}
+		if (!isStreamingMode && tableSource instanceof StreamTableSource &&
+			!((StreamTableSource<?>) tableSource).isBounded()) {
+			throw new ValidationException("Only bounded StreamTableSource can be used in batch mode.");
+		}
+
+		return new TableSourceTable<>(
+			relOptSchema,
+			names,
+			rowType,
+			statistic,
+			tableSource,
+			isStreamingMode,
+			table);
+	}
+
+	private static FlinkPreparingTableBase convertCatalogTable(
+			RelOptSchema relOptSchema,
+			List<String> names,
+			RelDataType rowType,
+			CatalogTable catalogTable,
+			CatalogSchemaTable schemaTable) {
+		TableSource<?> tableSource;
```

Comment:
Actually, after you introducing `CatalogSourceTable`, we can defer the logic of finding `TableSource` to `CatalogSourceTable.toRel`?

Reply:
We can, thanks.